src/server/ua_subscription_alarms_conditions.c

```diff
@@ -0,0 +1,2383 @@
+                            }
+                            return UA_STATUSCODE_GOOD;
+
+                        case UA_ENTERING_CONFIRMEDSTATE:
+                            if(conditionEntryTmp->specificCallbacksData.enteringConfirmedStateCallback != NULL) {
+                              *removeBranch = conditionEntryTmp->specificCallbacksData.confirmedRemoveBranch;
+                              return conditionEntryTmp->specificCallbacksData.enteringConfirmedStateCallback(server, condition);
+                            }
+                            return UA_STATUSCODE_GOOD;
+
+                        case UA_ENTERING_ACTIVESTATE:
+                            if(conditionEntryTmp->specificCallbacksData.enteringActiveStateCallback != NULL) {
+                              return conditionEntryTmp->specificCallbacksData.enteringActiveStateCallback(server, condition);
+                            }
+                            return UA_STATUSCODE_GOOD;
+
+                        default:
+                            return UA_STATUSCODE_BADNOTFOUND;
+                    }
+                }
+            }
+        }
+    }
+    return UA_STATUSCODE_BADNOTFOUND;
+}
+
+/**
+ * Gets the parent NodeId of a Field (e.g. Severity) or Field Property (e.g. EnabledState/Id)
+ */
+static UA_StatusCode
+getFieldParentNodeId(UA_Server *server, const UA_NodeId *field, UA_NodeId *parent) {
+    *parent = UA_NODEID_NULL;
+    UA_NodeId hasPropertyType = UA_NODEID_NUMERIC(0, UA_NS0ID_HASPROPERTY);
+    UA_NodeId hasComponentType = UA_NODEID_NUMERIC(0, UA_NS0ID_HASCOMPONENT);
+    const UA_VariableNode *fieldNode = (const UA_VariableNode *)UA_Nodestore_get(server, field);
+    if(fieldNode != NULL) {
+        for(size_t i = 0; i < fieldNode->referencesSize; i++) {
+            if((UA_NodeId_equal(&fieldNode->references[i].referenceTypeId, &hasPropertyType) ||
+               UA_NodeId_equal(&fieldNode->references[i].referenceTypeId, &hasComponentType)) &&
+               (true == fieldNode->references[i].isInverse)) {
+                UA_StatusCode retval = UA_NodeId_copy(&fieldNode->references[i].targetIds->nodeId, parent);
+                UA_Nodestore_release(server, (const UA_Node *)fieldNode);
+                return retval;
+            }
+        }
+    }
+
+    if(fieldNode != NULL)
+        UA_Nodestore_release(server, (const UA_Node *)fieldNode);
+
+    return UA_STATUSCODE_BADNOTFOUND;
+}
+
+/**
+ * Gets the NodeId of a Field (e.g. Severity)
+ */
+static UA_StatusCode
+getConditionFieldNodeId(UA_Server *server,
+                        const UA_NodeId *conditionNodeId,
+                        const UA_QualifiedName* fieldName,
+                        UA_NodeId *outFieldNodeId) {
+    UA_BrowsePathResult bpr = UA_Server_browseSimplifiedBrowsePath(server, *conditionNodeId, 1, fieldName);
+    if(bpr.statusCode != UA_STATUSCODE_GOOD || bpr.targetsSize < 1) {
```

Comment:
Can this return good if targetsSize == 0?

Reply:
I implemented it just like in
https://github.com/open62541/open62541/blob/d89e9e55bd810a1d1371dbb94dbe968d318bcdbf/src/server/ua_subscription_events.c#L345
without double checking , my bad...

Now that you mentioned it, I think it can't return good if targetsSize==0
https://github.com/open62541/open62541/blob/b13ad2edb2942939bd5a1aebdd6295c35a69d1a7/src/server/ua_services_view.c#L701